pallets/payment/src/types.rs

```diff
@@ -77,7 +86,11 @@ pub trait DisputeResolver<Account> {
 }
 
 /// Fee Handler trait that defines how to handle marketplace fees to every payment/swap
-pub trait FeeHandler<Account> {
+pub trait FeeHandler<Asset, Amount, Account> {
 	/// Get the distribution of fees to marketplace participants
-	fn apply_fees(from: &Account, to: &Account) -> (Account, Percent);
+	fn apply_fees(
+		from: &Account,
+		to: &Account,
+		detail: &PaymentDetail<Asset, Amount, Account>,
```

Comment:
mmm this is not gonna scale well if every time we change the payment details we have to change our trait. One last change could be to change payment detail to be generic over a single parameter i.e. `T` that is bound to be `Config` and then we can use `AssetIdOf<T>` and `BalanceOf<T>` inside `PaymentDetail<T>`

Reply:
Not sure if this is what you mean,

```rust
pub struct PaymentDetail<T : Config> {
	/// type of asset used for payment
	pub asset: AssetIdOf<T>,
	/// amount of asset used for payment
	pub amount: BalanceOf<T>,
	/// incentive amount that is credited to creator for resolving
	pub incentive_amount: BalanceOf<T>,
	/// enum to track payment lifecycle [Created, Released, Cancelled, NeedsReview]
	pub state: PaymentState,
	/// account that can settle any disputes created in the payment
	pub resolver_account: T::AccountId,
	/// fee charged and recipient account details
	pub fee_detail: Option<(T::AccountId, BalanceOf<T>)>,
	/// remarks to give context to payment
	pub remark: Option<Vec<u8>>, // TODO : switch to BoundedVec if possible
}
```

But trying this throws the error that TypeInfo cannot be implemented over Config

Reply:
> mmm this is not gonna scale well if every time we change the payment details we have to change our trait. One last change could be to change payment detail to be generic over a single parameter i.e. `T` that is bound to be `Config` and then we can use `AssetIdOf<T>` and `BalanceOf<T>` inside `PaymentDetail<T>`

Not sure if this is what you mean,

```rust
pub struct PaymentDetail<T : Config> {
	/// type of asset used for payment
	pub asset: AssetIdOf<T>,
	/// amount of asset used for payment
	pub amount: BalanceOf<T>,
	/// incentive amount that is credited to creator for resolving
	pub incentive_amount: BalanceOf<T>,
	/// enum to track payment lifecycle [Created, Released, Cancelled, NeedsReview]
	pub state: PaymentState,
	/// account that can settle any disputes created in the payment
	pub resolver_account: T::AccountId,
	/// fee charged and recipient account details
	pub fee_detail: Option<(T::AccountId, BalanceOf<T>)>,
	/// remarks to give context to payment
	pub remark: Option<Vec<u8>>, // TODO : switch to BoundedVec if possible
}
```

But trying this throws the error that TypeInfo cannot be implemented over Config